provider/oracle/images.go

```diff
@@ -0,0 +1,197 @@
+// For more information about oracle cloud shapes, please see:
+// https://docs.oracle.com/cloud/latest/stcomputecs/STCSA/api-Shapes.html
+// https://docs.oracle.com/cloud/latest/stcomputecs/STCSG/GUID-1DD0FA71-AC7B-461C-B8C1-14892725AA69.htm#OCSUG210
+func instanceTypes(c EnvironAPI) ([]instances.InstanceType, error) {
+	if c == nil {
+		return nil, errors.Errorf("cannot use nil client")
+	}
+
+	// fetch all shapes from the provider
+	shapes, err := c.AllShapes(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// convert shapes to InstanceType
+	onlyArch := []string{"amd64"}
+	types := make([]instances.InstanceType, len(shapes.Result), len(shapes.Result))
+	for key, val := range shapes.Result {
+		types[key].Name = val.Name
+		types[key].Arches = onlyArch
+		types[key].Mem = val.Ram
+		types[key].CpuCores = uint64(val.Cpus)
+		types[key].RootDisk = val.Root_disk_size
+	}
+
+	return types, nil
+}
+
+// findInstanceSpec returns an *InstanceSpec, imagelist name
+// satisfying the supplied instanceConstraint
+func findInstanceSpec(
+	c EnvironAPI,
+	allImageMetadata []*imagemetadata.ImageMetadata,
+	instanceType []instances.InstanceType,
+	ic *instances.InstanceConstraint,
+) (*instances.InstanceSpec, string, error) {
+
+	logger.Debugf("received %d image(s): %v", len(allImageMetadata), allImageMetadata)
+	version, err := series.SeriesVersion(ic.Series)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+	filtered := []*imagemetadata.ImageMetadata{}
+	for _, val := range allImageMetadata {
+		if val.Version != version {
+			continue
+		}
+		filtered = append(filtered, val)
+	}
+
+	images := instances.ImageMetadataToImages(filtered)
+	spec, err := instances.FindInstanceSpec(images, ic, instanceType)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	imagelist, err := getImageName(c, spec.Image.Id)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	return spec, imagelist, nil
+}
+
+func parseImageName(name string, uri *url.URL) (*imagemetadata.ImageMetadata, error) {
+	var id, arch, version string
+	if strings.HasPrefix(name, "Ubuntu") {
+		meta := strings.Split(name, ".")
+		if len(meta) < 4 {
+			return nil, errors.Errorf("invalid ubuntu image name: %s", name)
+		}
+		id = meta[len(meta)-1]
+		arch = meta[3]
+		version = meta[1] + "." + strings.TrimSuffix(meta[2], "-LTS")
+	} else if strings.HasPrefix(name, "Microsoft") {
+		if ver, ok := windowsServerMap[name]; ok {
+			version = ver
+			id = ver
+			arch = "amd64"
+		} else {
+			return nil, errors.Errorf("unknown windows version: %q", name)
+		}
+	} else {
+		return nil, errors.Errorf("could not determine OS from image name: %q", name)
+	}
+
+	tmp := strings.Split(uri.Host, ".")
+	region := tmp[0]
+	if len(tmp) > 1 {
+		region = tmp[1]
+	}
+	return &imagemetadata.ImageMetadata{
+		Id:         id,
+		Arch:       arch,
+		Endpoint:   fmt.Sprintf("%s://%s", uri.Scheme, uri.Host),
+		RegionName: region,
```

Comment:
AFAICT, the region name is pointless, and is only being used to match the image metadata we're creating.

I think we may as well drop the region name from the oracle cloud schema, and stop populating RegionName here and Region in the InstanceConstraint. That's one less thing the user needs to do to get things working.